File: crates/primitives/src/models/entities/bank.rs
use crate::models::entities::enum_types::CurrencyCode;
use chrono::{DateTime, Utc};
use diesel::{Associations, Identifiable, Insertable, Queryable};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;

#[derive(Queryable, Identifiable, Debug, Clone, Serialize, ToSchema, Deserialize)]
#[diesel(table_name = crate::schema::banks)]
pub struct Bank {
    pub id: i64,
    pub name: String,
    pub code: String,
    pub currency: CurrencyCode,
    pub country: String,
    pub is_active: bool,
}

#[derive(Insertable, Deserialize)]
#[diesel(table_name = crate::schema::banks)]
pub struct NewBank {
    pub id: i64,
    pub name: String,
    pub code: String,
    pub currency: CurrencyCode,
    pub country: String,
    pub is_active: bool,
}

#[derive(Debug, Queryable, Identifiable, Associations, ToSchema, Serialize)]
#[diesel(table_name = crate::schema::bank_accounts)]
#[diesel(belongs_to(crate::models::entities::user::User))]
pub struct BankAccount {
    pub id: Uuid,
    pub user_id: Uuid,
    pub bank_code: String,
    pub account_number: String,
    pub account_name: Option<String>,
    pub bank_name: Option<String>,
    pub provider_recipient_id: Option<String>,
    pub is_verified: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Insertable, Deserialize)]
#[diesel(table_name = crate::schema::bank_accounts)]
pub struct NewBankAccount<'a> {
    pub user_id: Uuid,
    pub bank_code: &'a str,
    pub account_number: &'a str,
    pub account_name: Option<&'a str>,
    pub bank_name: Option<&'a str>,
    pub provider_recipient_id: Option<&'a str>,
    pub is_verified: bool,
}

File: payego_ui/src/api/bank.ts
import client from './client';
import {BankAccount} from '@/types';

export interface Bank {
    code: string;
    name: string;
}

export const bankApi = {
    getBanks: () => client.get<{ banks: Bank[] }>('/api/banks/all').then(res => res.data.banks),
    getUserBankAccounts: () => client.get<{
        bank_accounts: BankAccount[]
    }>('/api/user/banks').then(res => res.data.bank_accounts),
    addBankAccount: (data: { account_number: string, bank_code: string, bank_name: string }) =>
        client.post('/api/banks/add', data),
    resolveAccount: (bankCode: string, accountNumber: string) =>
        client.get<{
            account_name: string
        }>(`/api/bank/resolve?account_number=${accountNumber}&bank_code=${bankCode}`).then(res => res.data),
    deleteBankAccount: (id: string) => client.delete(`/api/banks/${id}`),
};
